lib/Runtime/Library/Utf8String.h

```diff
@@ -0,0 +1,142 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#pragma once
+
+#include "Codex/Utf8Codex.h"
+
+namespace Js
+{
+    class Utf8String : public JavascriptString
+    {
+    private:
+        typedef struct {
+            FieldNoBarrier(size_t) length;
+            Field(char*) buffer;
+        } PrefixedUtf8String;
+        Field(PrefixedUtf8String*) utf8String;
+
+    protected:
+        DEFINE_VTABLE_CTOR(Utf8String, JavascriptString);
+
+    private:
+        
+        void SetUtf8Buffer(_In_reads_(utf8Length) char* buffer, size_t utf8Length)
+        {
+            this->utf8String = RecyclerNew(this->GetRecycler(), PrefixedUtf8String);
+            this->utf8String->length = utf8Length;
+            this->utf8String->buffer = buffer;
+        }
+
+    public:
+
+        Utf8String(_In_ JavascriptString* originalString, _In_reads_(utf8Length) char* buffer, size_t utf8Length, _In_ StaticType* type) :
+            JavascriptString(type),
+            utf8String(nullptr)
+        {
+            SetUtf8Buffer(buffer, utf8Length);
+
+            this->SetLength(originalString->GetLength());
+            this->SetBuffer(originalString->UnsafeGetBuffer());
```

Comment:
My thinking was that this should keep the finalized-ness of the source string: if the original string isn't finalized, it won't have a buffer so we would be setting it to nullptr (and then if someone asks for it, we generate it from the utf8 version), and otherwise it was finalized so we can remain finalized. I hadn't considered the case when a string would have an invalid buffer like a substring would.

Using `getSz` would be safe, but could maybe flatten more strings than expected... but it avoids having to decode the utf8 in future anyway. I think I will go with that.

Reply:
I'd be fine with something that said explicitly "if finalized, GetSz, else null", but UnsafeGetBuffer is marked unsafe for a reason :)

---
In reply to: [205623819](https://github.com/Microsoft/ChakraCore/pull/5348#discussion_r205623819) [](ancestors = 205623819)

Reply:
Actually even with GetSz the PropertyString problem persists. It returns an interior pointer to a PropertyRecord, which is kept alive by the string itself. We risk it getting collected from under us with this setup.

---
In reply to: [205624092](https://github.com/Microsoft/ChakraCore/pull/5348#discussion_r205624092) [](ancestors = 205624092,205623819)

Reply:
Buffer ownership is also a very real problem with the conversion code: we swap out a vtable, overwrite our owning pointer with some other data, and then our data might get collected.

---
In reply to: [205624399](https://github.com/Microsoft/ChakraCore/pull/5348#discussion_r205624399) [](ancestors = 205624399,205624092,205623819)